### DISCUSSION TITLE
transfer all balance

### DISCUSSION THREAD

**Discussion by alphered789375639** [Python-Cardano/pycardano#182] [Q&A]
Hello, I hope you are well.
I have been working on a project for some time
I want to transfer my entire account balance minus fee to another account using pycardano but I don't know how?
I tried some methods but I encountered various errors
Is this workable? If yes, how?
I've been trying for a long time and didn't get the result.. Please help

**Comment by cffls** [comment]
A similar question is asked here https://github.com/Python-Cardano/pycardano/discussions/123

**Reply by alphered789375639** [reply]
I can't find and import the wallet.py class.
I think that solution is very different from the one I'm looking for

**Comment by cffls** [comment]
You can do something like this:

```python
utxos = chain_context.utxos(str(current_address))
builder = TransactionBuilder(chain_context)

for utxo in utxos:
    builder.add_input(utxo)

signed_tx = builder.build_and_sign(signing_keys=[your_signing_key], change_address=new_address)
```

**Reply by alphered789375639** [reply]
I apologize for taking up your time..
I have some problems.
First, when I want to get the fee.
I want to create a transaction with the all balance and calculate the fee for that transaction.
I write this code but it doesn't work.

from pycardano import TransactionBuilder, BlockFrostChainContext, TransactionOutput, Address, SigningKey

API_KEY = 'MY_API_KEY'
URL = 'https://cardano-mainnet.blockfrost.io/api'
client = BlockFrostChainContext(API_KEY, base_url=URL)

from_address = 'MY_ADDRESS'
to_address = 'TO_ADDRESS'

balance = 1.064246

UTXOs = client.utxos(from_address)

transaction = TransactionBuilder(client).add_input_address(from_address).add_output(
    TransactionOutput.from_primitive([to_address, int(balance * 1_000_000)]))
signed_transaction = transaction.build_and_sign(['my_signing_key'],
                                                change_address=Address.from_primitive(from_address))
for utxo in UTXOs:
    transaction.add_input(utxo)
fee = transaction.build(Address.from_primitive(from_address)).fee / 1_000_000
print(fee)


How do I get the fee for a transaction whose value is the all balance of the address?

**Reply by cffls** [reply]
txbuilder will automatically calculate the fee for you in `build`. You don't need to specify transaction output in this case, because you can directly set change address to `to_address`. The following code will be sufficient I think:

```python
UTXOs = client.utxos(from_address)
transaction = TransactionBuilder(client)
for utxo in UTXOs:
    transaction.add_input(utxo)

signed_transaction = transaction.build_and_sign(['my_signing_key'],
change_address=Address.from_primitive(to_address))
print(signed_transaction)

# Now you can submit signed_transaction directly
...
```

**Reply by alphered789375639** [reply]
I tried your code and got this error, in this line (signed_transaction = transaction.build_and_sign(['my_signing_key'],
change_address=Address.from_primitive(to_address)))

I got this error:

Traceback (most recent call last):
File "/home/ADA/venv/lib/python3.10/site-packages/pycardano/coinselection.py", line 99, in select
    raise InsufficientUTxOBalanceException("UTxO Balance insufficient!")
pycardano.exception.InsufficientUTxOBalanceException: UTxO Balance insufficient!

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ADA/venv/lib/python3.10/site-packages/pycardano/txbuilder.py", line 1003, in build
    raise UTxOSelectionException(
pycardano.exception.UTxOSelectionException: All UTxO selectors failed.
Requested output:
 {'coin': 162157, 'multi_asset': {}} 
Pre-selected inputs:
 {'coin': 1064246, 'multi_asset': {}} 
Additional UTxO pool:
 [] 
Unfulfilled amount:
 {'coin': -902089, 'multi_asset': {}}

I have faced this error before

**Reply by cffls** [reply]
This looks like a bug to me. For me to reproduce it, could you provide the `from_address` in your script? Thanks!

**Reply by alphered789375639** [reply]
from_address = addr1qxn2q550aqh556p5e6mtf87vjrz6aat829qukchkr8na2y73rpplun0vgz2nsqmtkw3axz7zwd9a0vm8aurvacxu32sqgh7pw9

**Reply by alphered789375639** [reply]
I am still waiting for your answer. Thanks!

**Reply by cffls** [reply]
Thanks for the info. Looks like there is only `1.064246` ADA left in this account. The minimum output amount is 1 ADA, and the transaction fee is usually 0.16+. We can see that the amount of ADA sitting under this account is not able to afford min ADA + tx fee. Therefore, the UTXO selection failed. You can send one more ADA to this address to solve this problem.

**Reply by alphered789375639** [reply]
Thank you very much for your answers!
The problem was solved.

**Reply by cffls** [reply]
Thanks for confirming. Could you please kindly mark this as the answer? Thank you!
